File: archive/php/muse/data_structures/CircularlyLinkedListTest.php
<?php

declare(strict_types=1);

require_once "muse/data_structures/CircularlyLinkedList.php";
require_once "muse/util/TestRunner.php";

use muse\data_structures\CircularlyLinkedList;
use muse\util\TestRunner;

function testCircularlyLinkedList(): bool
{
    $size = 8192;

    $list = new CircularlyLinkedList();

    for ($i = 1; $i <= $size; $i++) {
        $list->append($i);
    }

    if ($list->size() != $size) {
        return false;
    }

    for ($i = 0; $i < $size; $i++) {
        if ($list->get($i) != $i + 1) {
            return false;
        }
    }

    for ($i = 0; $i < $size; $i++) {
        $list->set($i, $size - $i);
    }

    for ($i = 0; $i < $size; $i++) {
        if ($list->get($i) != $size - $i) {
            return false;
        }
    }

    for ($i = 0, $j = $size - 1; $i < $j; $i++, $j--) {
        $x = $list->get($i);
        $y = $list->get($j);

        $list->remove($i);
        $list->insert($i, $y);
        $list->remove($j);
        $list->insert($j, $x);
    }

    for ($i = 0; $i < $size; $i++) {
        if ($list->get($i) != $i + 1) {
            return false;
        }
    }

    for ($i = $size; $i >= 1; $i--) {
        if ($list->back() != $i) {
            return false;
        }
        $list->eject();
    }

    return $list->isEmpty();
}

if (count(debug_backtrace()) == 0) {
    TestRunner\pick("testCircularlyLinkedList");
}

File: go/muse/data_structures/CircularlyLinkedList/CircularlyLinkedList_test.go
package CircularlyLinkedList

import "testing"

func TestCircularlyLinkedList(t *testing.T) {
    size := 8192

    list := New()

    for i := 1; i <= size; i++ {
        list.Append(i)
    }

    if list.Size() != size {
        t.FailNow()
    }

    for i := 0; i < size; i++ {
        if list.Get(i) != i + 1 {
            t.FailNow()
        }
    }

    for i := 0; i < size; i++ {
        list.Set(i, size - i)
    }

    for i := 0; i < size; i++ {
        if list.Get(i) != size - i {
            t.FailNow()
        }
    }

    var x int
    var y int

    for i, j := 0, size - 1; i < j; i, j = i + 1, j - 1 {
        x = list.Get(i)
        y = list.Get(j)

        list.Remove(i)
        list.Insert(i, y)
        list.Remove(j)
        list.Insert(j, x)
    }

    for i := 0; i < size; i++ {
        if list.Get(i) != i + 1 {
            t.FailNow()
        }
    }

    for i := size; i >= 1; i-- {
        if list.Back() != i {
            t.FailNow()
        }
        list.Eject()
    }

    if !list.IsEmpty() {
        t.FailNow()
    }
}
